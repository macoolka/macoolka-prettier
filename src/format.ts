/**
 * code formatter
 * @desczh
 * 格式化代码
 * @file
 */
import * as prettier from 'prettier';
import { Options, BuiltInParserName } from 'prettier';
import { isString, isEmpty } from 'macoolka-predicate';
import { omit, merge } from 'macoolka-object';
import { MonadFunction, fromReader } from 'macoolka-app/lib/MonadFunction';
const formatContent = (content: string | object): string =>
    isString(content)
        ? content
        : JSON.stringify(content);

const getParserName = (extname: string): BuiltInParserName => {
    switch (extname) {
        case 'json':
            return 'json';
        case 'markdown':
            return 'markdown';
        case 'graphql':
            return 'graphql';
        case 'mdx':
            return 'mdx';
        case 'css':
            return 'css';
        case 'html':
            return 'html';
        default:
            return 'typescript';
    }
};
export interface PrettierInput extends Options {
    extname?: string;
    content: string | object;
}
/**
 * Format a code
 * @desczh
 * 格式化代码
 * @since 0.2.0
 */
export const format = (a: PrettierInput): string => {
    const extname = isEmpty(a.extname) ? 'ts' : a.extname;
    const parse = isEmpty(a.parser) ? getParserName(extname) : a.parser;
    const option = merge({}, {
        semi: false,
        singleQuote: true,
        printWidth: 120,
        parse,
    }, omit(a, ['content', 'extname']));

    return prettier.format(formatContent(a.content), option);
};
/**
 * Format a code
 * @desczh
 * 格式化代码
 * @since 0.2.5
 */
export const formatString = (o: Options) => (a: string|object): string => {

    const option = merge({}, {
        semi: false,
        singleQuote: true,
        printWidth: 120,
    }, o);

    return prettier.format(formatContent(a), option);
};
/**
 * @since 0.2.5
 */
export const formatTs = formatString({ parser: 'typescript' })
/**
 * @since 0.2.5
 */
export const formatCss = formatString({ parser: 'css' })
/**
 * @since 0.2.5
 */
export const formatJson = formatString({ parser: 'json' })
/**
 * @since 0.2.5
 */
export const formatMarkdown = formatString({ parser: 'markdown' })
/**
 * @since 0.2.5
 */
export const formatGraphql = formatString({ parser: 'graphql' })
/**
 * @since 0.2.5
 */
export const formatMdx = formatString({ parser: 'mdx' })
/**
 * @since 0.2.5
 */
export const formatHtml = formatString({ parser: 'html' })
/**
 * @since 0.2.5
 */
export default formatString
/**
 * Format a code Monad
 * @desczh
 * 格式化代码(Monad)
 * @since 0.2.5
 */
export const monadFormat = (o: Options): MonadFunction< string|object, string> => fromReader(formatString(o));
