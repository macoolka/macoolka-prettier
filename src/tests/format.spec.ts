import format,{ monadFormat,formatCss,formatTs } from '../format'
import { resolve } from 'path'
import { readFileSync } from 'fs'
import { pipe } from 'fp-ts/lib/pipeable'
import * as E from 'fp-ts/lib/Either'
import * as MN from 'macoolka-app/lib/MonadNode'
describe('format', () => {
    
    it('typescript', () => {
        const tsx = readFileSync(resolve(__dirname, 'formats', 'Modal.tsx'), { encoding: 'utf8' });
        expect(format({ parser: 'typescript' })(tsx)).toMatchSnapshot()

        expect(formatTs(tsx)).toMatchSnapshot()
    });
    it('css', () => {
        const css = `
        .a{
            color:red;
            margin:6;
        }
        `;
        expect(format({ parser: 'css' })(css)).toMatchSnapshot()

        expect(formatCss(css)).toMatchSnapshot()
    });
    it('object', () => {
        const json = {
            a: 1,
            b: 2,
            scripts: {
                prepublishOnly: "npm run clean && npm run reinstall && npm run lint  && npm run test && npm run build"
            }
        }

        expect(format({ parser: 'json'})(json)).toMatchSnapshot()


    });
    it('json', () => {
        const json = `{
            "scripts": {
                         "prepush": "npm run test",
"prepublishOnly": "npm run clean && npm run reinstall && npm run lint  && npm run test && npm run build",
        "coverage": "rimraf coverage && jest --config jest.config.json --coverage"
              },

        }
 `
        expect(format({ parser: 'json'})(json)).toMatchSnapshot()

    });


});

describe('monadFormat', () => {

    it('typescript', async () => {
        const tsx = readFileSync(resolve(__dirname, 'formats', 'Modal.tsx'), { encoding: 'utf8' });
        const result = await pipe(
            monadFormat({ parser: 'typescript' })(tsx),
            MN.map(a => {
                expect(a).toMatchSnapshot()
            }),
        )()
        expect(E.isRight(result))


    });
    it('object', async () => {
        const json = {
            a: 1,
            b: 2,
            scripts: {
                prepublishOnly: "npm run clean && npm run reinstall && npm run lint  && npm run test && npm run build"
            }
        }
        const result = await pipe(
            monadFormat({ parser: 'json'})(json),
            MN.map(a => {
                expect(a).toMatchSnapshot()
            }),
        )()
        expect(E.isRight(result))



    });
    it('json', async () => {
        const json = `{
            "scripts": {
                         "prepush": "npm run test",
"prepublishOnly": "npm run clean && npm run reinstall && npm run lint  && npm run test && npm run build",
        "coverage": "rimraf coverage && jest --config jest.config.json --coverage"
              },

        }
 `
        const result = await pipe(
            monadFormat({ parser: 'json'})(json),
            MN.map(a => {
                expect(a).toMatchSnapshot()
            }),
        )()
        expect(E.isRight(result))

    });
    it('json error', async () => {
        const json = `{
            "scripts": 
                         "prepush": "npm run test",
"prepublishOnly": "npm run clean && npm run reinstall && npm run lint  && npm run test && npm run build",
        "coverage": "rimraf coverage && jest --config jest.config.json --coverage"
              },

        }
 `
        const result = await pipe(
            monadFormat({ parser: 'json'})(json),
            MN.mapLeft(a => {
                expect(a({})).toMatchSnapshot()
            }),
        )()
        expect(E.isLeft(result))

    });

});

