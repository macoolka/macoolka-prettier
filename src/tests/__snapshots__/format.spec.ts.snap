// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format css 1`] = `
".a {
  color: red;
  margin: 6;
}
"
`;

exports[`format css 2`] = `
".a {
  color: red;
  margin: 6;
}
"
`;

exports[`format json 1`] = `
"{
  \\"scripts\\": {
    \\"prepush\\": \\"npm run test\\",
    \\"prepublishOnly\\": \\"npm run clean && npm run reinstall && npm run lint  && npm run test && npm run build\\",
    \\"coverage\\": \\"rimraf coverage && jest --config jest.config.json --coverage\\"
  }
}
"
`;

exports[`format object 1`] = `
"{
  \\"a\\": 1,
  \\"b\\": 2,
  \\"scripts\\": {
    \\"prepublishOnly\\": \\"npm run clean && npm run reinstall && npm run lint  && npm run test && npm run build\\"
  }
}
"
`;

exports[`format typescript 1`] = `
"export function rand() {
  return Math.round(Math.random() * 20) - 10
}

export function getModalStyle() {
  const top = 50 + rand()
  const left = 50 + rand()

  return {
    top: \`\${top}%\`,
    left: \`\${left}%\`,
    transform: \`translate(-\${top}%, -\${left}%)\`
  }
}
"
`;

exports[`format typescript 2`] = `
"export function rand() {
  return Math.round(Math.random() * 20) - 10
}

export function getModalStyle() {
  const top = 50 + rand()
  const left = 50 + rand()

  return {
    top: \`\${top}%\`,
    left: \`\${left}%\`,
    transform: \`translate(-\${top}%, -\${left}%)\`
  }
}
"
`;

exports[`monadFormat json 1`] = `
"{
  \\"scripts\\": {
    \\"prepush\\": \\"npm run test\\",
    \\"prepublishOnly\\": \\"npm run clean && npm run reinstall && npm run lint  && npm run test && npm run build\\",
    \\"coverage\\": \\"rimraf coverage && jest --config jest.config.json --coverage\\"
  }
}
"
`;

exports[`monadFormat json error 1`] = `"\\"SyntaxError: Unexpected token, expected \\\\\\",\\\\\\" (3:35)\\\\n\\\\u001b[0m \\\\u001b[90m 1 | \\\\u001b[39m{\\\\u001b[0m\\\\n\\\\u001b[0m \\\\u001b[90m 2 | \\\\u001b[39m            \\\\u001b[32m\\\\\\"scripts\\\\\\"\\\\u001b[39m\\\\u001b[33m:\\\\u001b[39m \\\\u001b[0m\\\\n\\\\u001b[0m\\\\u001b[31m\\\\u001b[1m>\\\\u001b[22m\\\\u001b[39m\\\\u001b[90m 3 | \\\\u001b[39m                         \\\\u001b[32m\\\\\\"prepush\\\\\\"\\\\u001b[39m\\\\u001b[33m:\\\\u001b[39m \\\\u001b[32m\\\\\\"npm run test\\\\\\"\\\\u001b[39m\\\\u001b[33m,\\\\u001b[39m\\\\u001b[0m\\\\n\\\\u001b[0m \\\\u001b[90m   | \\\\u001b[39m                                  \\\\u001b[31m\\\\u001b[1m^\\\\u001b[22m\\\\u001b[39m\\\\u001b[0m\\\\n\\\\u001b[0m \\\\u001b[90m 4 | \\\\u001b[39m\\\\u001b[32m\\\\\\"prepublishOnly\\\\\\"\\\\u001b[39m\\\\u001b[33m:\\\\u001b[39m \\\\u001b[32m\\\\\\"npm run clean && npm run reinstall && npm run lint  && npm run test && npm run build\\\\\\"\\\\u001b[39m\\\\u001b[33m,\\\\u001b[39m\\\\u001b[0m\\\\n\\\\u001b[0m \\\\u001b[90m 5 | \\\\u001b[39m        \\\\u001b[32m\\\\\\"coverage\\\\\\"\\\\u001b[39m\\\\u001b[33m:\\\\u001b[39m \\\\u001b[32m\\\\\\"rimraf coverage && jest --config jest.config.json --coverage\\\\\\"\\\\u001b[39m\\\\u001b[0m\\\\n\\\\u001b[0m \\\\u001b[90m 6 | \\\\u001b[39m              }\\\\u001b[33m,\\\\u001b[39m\\\\u001b[0m\\""`;

exports[`monadFormat object 1`] = `
"{
  \\"a\\": 1,
  \\"b\\": 2,
  \\"scripts\\": {
    \\"prepublishOnly\\": \\"npm run clean && npm run reinstall && npm run lint  && npm run test && npm run build\\"
  }
}
"
`;

exports[`monadFormat typescript 1`] = `
"export function rand() {
  return Math.round(Math.random() * 20) - 10
}

export function getModalStyle() {
  const top = 50 + rand()
  const left = 50 + rand()

  return {
    top: \`\${top}%\`,
    left: \`\${left}%\`,
    transform: \`translate(-\${top}%, -\${left}%)\`
  }
}
"
`;
